### ISSUE TITLE
Docker containers get incorrect HTTP_PROXY injected if the host has HTTP_PROXY=http://127.0.0.1:xxx/

### ISSUE THREAD

**Issue by brekelj1** [actions/runner#868]
**Describe the bug**
Running the runner in an environment such that the HTTP proxy runs on the same host as the runner, and the HTTP_PROXY et all environment variables are set to `http://127.0.0.1:xxx/`, results in [docker containers being configured](https://github.com/actions/runner/blob/7ff6ff6afa8b88434cfb5e766b29e5fb9f2bb19a/src/Runner.Worker/Container/ContainerInfo.cs#L245) to have their HTTP_PROXY et all environment variables set to `http://127.0.0.1:xxx/`. But docker containers by default run in the default bridge network, so connecting to `127.0.0.1:xxx` from container basically always results in connection refused errors.

You could argue running the HTTP proxy on the same host is a dumb idea, because it means that docker containers cannot access the HTTP proxy, thereby breaking any docker actions that need to use the proxy. But in my environment of self-hosted runners we do the following: run a local HTTP forward proxy (that forwards to a parent proxy) as a docker container with a port published to the host, and set HTTP_PROXY et all environment variables of any other docker containers (on the default bridge network) by configuring `$HOME/.docker/config.json` (see [docs](https://docs.docker.com/network/proxy/#configure-the-docker-client)). This way we can centrally configure credentials for the parent proxy and we can centrally configure a proxy bypass list. Without the central HTTP forward proxy we would require that any possible tools running in jobs:
1. Support the authentication scheme to the parent proxy.
2. Consistently interpret the proxy bypass list (but many tools don't support CIDR notation, for example). Since our local HTTP forward proxy is written in Go we support CIDR notation in the proxy bypass list.

Running the proxy not on the same host, but on another compute unit, is not an ideal alternative because of 1 (authentication to the proxy).

It seems that docker container proxy configuration configured in `$HOME/.docker/config.json` (see [docs](https://docs.docker.com/network/proxy/#configure-the-docker-client)) is always overridden.

Recently many workflows on self-hosted runners broke because they relied on pre https://github.com/actions/runner/pull/840 behavior.

**To Reproduce**
1. Run a runner with a HTTP proxy set to `http://127.0.0.1:xxx`.
2. Run a step with `uses: docker://` syntax that needs to use the proxy.
3. Observe that the container of the step gets a connection refused error.

**Expected behavior**
If the runner's HTTP_PROXY et all environment variables are set to `localhost` or `127.0.0.1` (and/or if `$HOME/.docker/config.json`, [docs](https://docs.docker.com/network/proxy/#configure-the-docker-client) is configured then docker containers do not explicitly get `HTTP_PROXY` set. 

## Runner Version and Platform
Version of your runner?
https://github.com/actions/runner/releases/tag/v2.275.1

OS of the machine running the runner? OSX/Windows/Linux/...
Linux

## What's not working?
A step with `uses: docker://` such that the container does an HTTP request via the HTTP proxy configured in `HTTP_PROXY` gets a connection refused error because `HTTP_PROXY` is set to `127.0.0.1`.

**Comment by klassiker** [comment]
Sad to see this issue is still not fixed. I thought they would just revert my pull request and be done with it.

I don't have the time to investigate and fix this, but this would be a the solution:

- revert my pull request
- fix the environment variables for executing the docker command since this is the step where the `~/.docker/config.json` gets ignored

Result: separate proxy configuration for the runner and the container.

When I did my first fix for the issue I noticed this was the original problem but after trying to understand this huge undocumented code base I somehow got lost and did something else. Sorry for that.

I hope someone reads this and is willing to fix it...

**Comment by klassiker** [comment]
Oh, forgot to mention a possible workaround:

could you please try hosting the runner inside it's own docker container?
This way it's located on the same host and the environment variables for the runner are the ones that are also correct for the docker container spawned by the action.

Not sure if that will work and your workflows are probably edited by now - but who knows.

**Comment by sodaws** [comment]
Hi,
i'm in a similar situation and would like to propose another workaround in addition to @klassiker's and a description of how this behaviour occurs and why it is still a problem even with the workaround.

##  tl;dr
Adjust your workflow like this:

```

jobs:
  test-container-env:
    runs-on: self-hosted
    container:
      image: curlimages/curl:latest
      env:
        http_proxy: http://host.docker.internal:xxxx
        https_proxy: http://host.docker.internal:xxxx
      options: --add-host=host.docker.internal:host-gateway
    steps:
    - name: curl google
      run: |
        export HTTP_PROXY=http://host.docker.internal:xxxx
        export HTTPS_PROXY=http://host.docker.internal:xxxx
        curl google.com
    # repeat the export every single run command
     - name: install random package
       run: |
         export HTTP_PROXY=http://host.docker.internal:xxxx
         export HTTPS_PROXY=http://host.docker.internal:xxxx
         pip install pycodestyle
```

 ## My environment
similar to @brekelj1's:
- a Linux system
- a host that can connect to the internet only through a proxy
- a proxy forwarder on localhost:xxxx
- Docker, configured to use the proxy on localhost to connect to dockerhub
- a self-hosted gitlab-runner on the same host that can connect to github through the proxy by setting `http_proxy=http://localhost:xxxx` and `https_proxy=http://localhost:xxxx` in its .env file
- only difference: i'm not using the `~/.docker/config.json` to configure the proxy for the containers started by the runner

## Problem description

### Unexpected behaviour when setting the proxy in the .env file
Ok, so the behaviour of the github runner is a little bit confusing in this scenario and I'm basically just gonna repeat what @brekelj1 wrote but in other words.

If you set `http_proxy` and `https_proxy` (lowercase) or `HTTP_PROXY` and `HTTPS_PROXY` (uppercase) in the .env file of the runner, the following happens.

If a job is picked up, the runner will not only apply your defined http(s)_proxy urls but also duplicates them, so that you always end up with four environment variables instead of two.

```
http_proxy=http://localhost:xxxx
https_proxy=http://localhost:xxxx
HTTP_PROXY=http://localhost:xxxx
HTTPS_PROXY=http://localhost:xxxx
```

If your .env contains lowercase variable names, they will be copied to uppercase and the other way around. **This is the first unexpected and (debatable) unwanted behaviour.**

### This does not happen to variables with other names
This does not happen with ANY (at least not with the ones that I tested) other variables that you define in the .env file. This means that `test=123` in the .env will also just be `test=123` in the runner environment. By runner environment I mean the output of a worflow like this:

``` 
name: print runner environment variables
on: workflow_dispatch
jobs:
  test-runner-env:
    runs-on: ubuntu-latest 
    # self-hosted in my case 
    steps:
    - name: print env
      run: printenv
```

The environment variable `test=123` will not appear in docker containers started by the runner just like you would expect. To make them available in the container you would need to configure your workflow like this:

```
name: print container environment variables
on: workflow_dispatch
jobs:
  test-container-env:
    runs-on: ubuntu-latest
    container:
      image: curlimages/curl:latest
      env:
        test: 123
    steps:
    - name: print env
      run: printenv
```

The definition in the .env file doesn't matter in this case. It will only be available in the runner's "own" environment and not in the docker containers.

### The second unexpected/unwanted behaviour
The http(s)_proxy environment variables of the .env file are automatically inserted into all of the containers that the runner starts. If you have a workflow like the one in the second example just above, all four proxy environment variables are inserted into the container (uppercase and lowercase http(s)_proxy=...).

Now this is a little bit of a problem because in this scenario, we don't want our applications inside the containers to connect to `http://localhost:xxxx` because that would mean that they connect to the **container internal localhost**. We want to connect to the localhost on the host of our runner so that our proxy forwarder can do its job and redirect the requests to the actual proxy server which connects us to the internet.

This means, what we actually want is our docker container connecting to `http://host.docker.internal:xxxx` . This will (if I understood correctly) foward the traffic from inside the container to our docker0 interface on the host with the ip 172.17.0.x from where the proxy forwarder can pick it up and then redirects us to our actual proxy and to the internet.

How do we do this? We change the http(s)_proxy environment variables of the container to `http://host.docker.internal:xxxx` (xxxx being the port of your proxy forwarder). This works only on somewhat newer versions of Docker (at least since 18.x) and the URL might be different on Mac and Windows, I'm using Linux.

In our workflow file this would look like this:

```
name: print container environment variables
on: workflow_dispatch
jobs:
  test-container-env:
    runs-on: self-hosted
    container:
      image: curlimages/curl:latest
      env:
        test: 123
        http_proxy: http://host.docker.internal:xxxx
        https_proxy: http://host.docker.internal:xxxx
      options: --add-host=host.docker.internal:host-gateway
    steps:
    - name: print env
      run: printenv
    - name: curl google
      run: curl google.com
```

If we run this workflow, everything works fine!

### But there still is a problem
Remember how the proxy environment variables are loaded? That's the problem. With the workflow above, your container environment will have these variables:

```
http_proxy=http://host.docker.internal:xxxx
https_proxy=http://host.docker.internal:xxxx
HTTP_PROXY=http://localhost:xxxx
HTTPS_PROXY=http://localhost:xxxx
```

"It works, why is that a problem, just add them in your workflow definition?"

Well, exactly that does not work.

### The third unexpected/unwanted behaviour

If we just change the workflow.yml, just like what you would think the solution would be:

```
name: print container environment variables
on: workflow_dispatch
jobs:
  test-container-env:
    runs-on: self-hosted
    container:
      image: curlimages/curl:latest
      env:
        test: 123
        http_proxy: http://host.docker.internal:xxxx
        https_proxy: http://host.docker.internal:xxxx
        HTTP_PROXY: http://host.docker.internal:xxxx
        HTTPS_PROXY: http://host.docker.internal:xxxx
      options: --add-host=host.docker.internal:host-gateway
    steps:
    - name: print env
      run: printenv
    - name: curl google
      run: curl google.com
```

The Github syntax checker for workflows will throw [this ](https://i.imgur.com/9OKSkAn.png) error:

`This workflow is not valid ... HTTP_PROXY is already defined ... HTTPS_PROXY is already defined`

The github runner [documentation ](https://docs.github.com/en/actions/reference/environment-variables#about-environment-variables) says: `Environment variables are case-sensitive.` 

Well, actually they are kind of not, I guess?! They are respected if I define them in the .env of the runner but not in any other case. The syntax checker will not even let me start the workflow. The runner and docker can handle case-sensitive variables but I can't even use/define them because the syntax checker says that a workflow with the same two variables (one uppercase and one lowercase) is invalid.

### Having two different proxy URLs can cause problems

Unfortunately, I can't live with this situation where one pair of my proxy environment variables in my container contains the wrong proxy URLs. It's always a gamble which one any given program in my container will read (first). Will it be the uppercase or the lowercase ones? If it's the wrong one, the container will try to connect to its internal localhost and im getting a `connection refused` error. 

### The workaround

There is a non-optimal solution to this. Firstly, I was just trying to fix this by adding these lines before my other run commands:
```
- name: override wrong env variables
  run: |
    export HTTP_PROXY=http://host.docker.internal:xxxx
    export HTTP_PROXY=http://host.docker.internal:xxxx
- name: do stuff
  run: install some package from the internet
```

This approach does not work because the environment variables are reset between the `run:` steps.

What you have to do is to include the export in every single run command where the wrong variable could possibly be read. This means that with the following approach, even the wrong proxy urls that the github syntax checker not allows us to override anywhere else will contain the correct URLs.

```
name: print container environment variables
on: workflow_dispatch
jobs:
  test-container-env:
    runs-on: self-hosted
    container:
      image: curlimages/curl:latest
      env:
        http_proxy: http://host.docker.internal:xxxx
        https_proxy: http://host.docker.internal:xxxx
      options: --add-host=host.docker.internal:host-gateway
    steps:
    - name: curl google
      run: |
        export HTTP_PROXY=http://host.docker.internal:xxxx
        export HTTPS_PROXY=http://host.docker.internal:xxxx
        curl google.com
        printenv
```

With this configuration, the environment variables in this step will be:

```
http_proxy=http://host.docker.internal:xxxx
https_proxy=http://host.docker.internal:xxxx
HTTP_PROXY=http://host.docker.internal:xxxx
HTTPS_PROXY=http://host.docker.internal:xxxx
```

Just like we wanted it! 😃

### Possible fixes

Right now, I can only think of four possible solutions for this problem:

1. Use @klassiker's workaround and host your runner inside a container. The downside of this approach is that you can't use docker-in-docker commands. `docker build` for example won't work because you already are inside a container.
2. Use my workaround
3. Make the github workflow syntax checker recognize the difference between lowercase and uppercase variables with the same name (e.g. http_proxy and HTTP_PROXY)
4. Remove the automatic insertion of the four http(s)_proxy variables into the containers in the `docker create` command that the runner executes to start the containers.

Can you replicate and verify this?

I can't tell if the behaviour of the syntax checker is intended or not. If it is not, we should open another issue about this...but i'm not sure if this repo is the right place to do that. If you do, link them to this issue to save yourself the typing 😄 

If you need more information, please let me know and please correct me this post contains any wrong statements.

**Comment by awsaba** [comment]
This is a huge problem for us when running services in an environment where proxies are required and all the relevant variables can not be overridden with one of the workarounds.

```
services:
   postgresql:
     image: posgresql:9.6
  other-service:
    image: someimage:latest
    env:
      NO_PROXY: postgresql
      HTTP_PROXY: ""
      HTTPS_PROXY: ""
```

The docker container for `other-service` is still created with the system values for (lowercase) `http_proxy`/`https_proxy`/`no_proxy` in addition to the ones in the `env`.

**Comment by richard-bain** [comment]
We have a work around for this issue. The following assumes your runners are linux based and your proxy is running in a container:

1. Find the gateway IP for your host: `docker network inspect bridge -f '{{range .IPAM.Config}}{{.Gateway}}{{end}}')`
2. Use that in your http/s proxy env variables on the host
3. Ensure this gateway ip is in `no_proxy`

This works because from the host's perspective this is a loopback to itself e.g. `172.17.0.1:3128` and a container on both the default bridge network and also any user-defined bridge network can also access this address.

Going one step further. if you are able to agree to a convention with your users you could do something like add `no_proxy=local,...` and remove the `.local` domain from the proxy. If everyone is prepared to do something like `docker run ... --network-alias=<name>.local` then this no proxy would also allow name based user-defined networking to work without having to modify the proxy every time.

FWIW I raised an upstream [issue on libnetwork](https://github.com/moby/libnetwork/issues/2645) but don't really expect action.

**Comment by brekelj1** [comment]
Closing due to the elegant workaround discovered by @richard-bain
